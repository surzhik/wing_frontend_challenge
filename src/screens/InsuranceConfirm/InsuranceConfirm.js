import React from "react";
import * as PropTypes from "prop-types";
import * as routes from "app/routes";
import Back from "common/Back";
import SubscriptionSwitcher from "subscriptions/SubscriptionSwitcher";
import SelectedPlan from "insurance/insurancePlans/SelectedPlan";
import styles from "../Insurance/Insurance.module.css";
import { connect } from "react-redux";

class InsuranceConfirm extends React.Component {
  state = {
    activating: false
  };

  confirmContract = async () => {
    const { subId, onActivationSuccess, showInfoMessage } = this.props;
    this.setState({
      activating: true
    });
    const contract = await this.createContract();

    if (contract && contract.id) {
      const device = await this.createDevice(contract.id);
      if (device && device.id) {
        const activate = await this.activateContract(contract.id);
        if (activate === "success") {
          showInfoMessage(
            "general",
            "Congratulations! Your contract successfully created."
          );
          onActivationSuccess(subId);
        }
      }
    }
    /*
    Activate submit on error
    this.setState({
      activating: false
    });
     */
  };

  createContract = async () => {
    const { createInsuranceContract, subId, subscription } = this.props;
    try {
      return await createInsuranceContract({
        [subscription]: subId
      });
    } catch (err) {
      //show something on error
      return err;
    }
  };

  createDevice = async contract => {
    const { createInsuranceDevice, sku, insPlanId } = this.props;
    try {
      return await createInsuranceDevice({
        contract,
        device_specs: sku,
        plan_type: insPlanId
      });
    } catch (err) {
      //show something on error
      return err;
    }
  };

  activateContract = async id => {
    const { activateInsuranceContract } = this.props;
    try {
      return await activateInsuranceContract({ id });
    } catch (err) {
      //show something on error
      return err;
    }
  };

  render() {
    const {
      subId,
      sku,
      insPlanId,
      fetchInsurancePlan,
      subscription
    } = this.props;
    const { activating } = this.state;
    return (
      <div>
        <SubscriptionSwitcher
          {...(subscription === "att_subscription"
            ? { attSubId: subId }
            : { sprintSubId: subId })}
          attRoute={routes.attSubscription}
          sprintRoute={routes.sprintSubscription}
        />
        <div className={styles.Back}>
          <Back to={routes.attInsurancePlan(subId, sku)} />
        </div>
        <h1>Confirm an insurance plan</h1>
        <SelectedPlan
          insurancePlanId={insPlanId}
          findInsurancePlan={fetchInsurancePlan}
          onConfirm={this.confirmContract}
          activating={activating}
        />
      </div>
    );
  }
}

InsuranceConfirm.propTypes = {
  subId: PropTypes.any,
  insPlanId: PropTypes.any,
  subscription: PropTypes.string,
  fetchInsurancePlan: PropTypes.func,
  activateInsuranceContract: PropTypes.func,
  createInsuranceContract: PropTypes.func,
  createInsuranceDevice: PropTypes.func,
  showInfoMessage: PropTypes.func,
  onActivationSuccess: PropTypes.func
};

export default InsuranceConfirm;
