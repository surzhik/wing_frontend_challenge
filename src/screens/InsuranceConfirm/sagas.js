import { put, call } from "redux-saga/effects";
import { setRedirect } from "auth/actions";
import { authenticateUser } from "auth/sagas";
import { attInsuranceConfirm } from "app/routes";

export function* sprintInsuranceConfirmNavigate() {
  yield call(authenticateUser);
}

export function* attInsuranceConfirmNavigate({ subId, sku, insPlanId }) {
  yield put(setRedirect(attInsuranceConfirm(subId, sku, insPlanId)));
  yield call(authenticateUser);
}
