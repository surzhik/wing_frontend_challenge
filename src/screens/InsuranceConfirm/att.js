import { connect } from "react-redux";
import * as routes from "app/routes";
import InsuranceConfirm from "./InsuranceConfirm";
import { find as fetchInsurancePlan } from "insurance/insurancePlans/actions";
import {
  createContract as createInsuranceContract,
  activateContract as activateInsuranceContract
} from "insurance/insuranceContracts/actions";
import { createDevice as createInsuranceDevice } from "insurance/insuredDevices/actions";
import { info as showInfoMessage } from "notifications/actions";

const mapStateToProps = state => ({
  subscription: "att_subscription"
});

const mapDispatchToProps = {
  fetchInsurancePlan,
  activateInsuranceContract,
  createInsuranceContract,
  createInsuranceDevice,
  showInfoMessage,
  onActivationSuccess: routes.attSubscription
};

export default connect(mapStateToProps, mapDispatchToProps)(InsuranceConfirm);
