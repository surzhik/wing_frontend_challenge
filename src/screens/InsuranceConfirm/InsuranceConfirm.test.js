import React from "react";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import { shallow } from "enzyme";
import InsuranceConfirm from "./InsuranceConfirm";
import SubscriptionSwitcher from "subscriptions/SubscriptionSwitcher";
import SelectedPlan from "insurance/insurancePlans/SelectedPlan";
const mockStore = configureMockStore();

describe("Insurance Confirmation", () => {
  const props = {
    subscription: "subscription",
    fetchInsurancePlan: jest.fn(),
    activateInsuranceContract: jest.fn(),
    createInsuranceContract: jest.fn(),
    createInsuranceDevice: jest.fn(),
    showInfoMessage: jest.fn(),
    onActivationSuccess: jest.fn()
  };

  describe("Insurance Confirmation structure", () => {
    it("have unchanged structure", () => {
      const store = mockStore({});
      const wrapper = shallow(
        <Provider store={store}>
          <InsuranceConfirm {...props} />
        </Provider>
      ).dive();
      expect(wrapper).toMatchSnapshot();
    });
  });

  const subscription = shallow(<InsuranceConfirm {...props} />);
  it("should render", () => {
    expect(subscription).toHaveLength(1);
  });

  it("should have selector", () => {
    expect(subscription.find(SubscriptionSwitcher)).toHaveLength(1);
  });
  it("should have 1 boxes with current props", () => {
    expect(subscription.find(SelectedPlan)).toHaveLength(1);
  });
});
