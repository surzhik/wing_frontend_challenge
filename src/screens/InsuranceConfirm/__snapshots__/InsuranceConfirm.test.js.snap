// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Insurance Confirmation Insurance Confirmation structure have unchanged structure 1`] = `
<div>
  <Connect(SubscriptionDecorator)
    attRoute={[Function]}
    sprintRoute={[Function]}
  />
  <div
    className="Back"
  >
    <Back
      to={
        Object {
          "id": "ATT_INSURANCE_PLAN",
          "params": Object {
            "sku": undefined,
            "subId": undefined,
          },
          "path": "/a/subscriptions/undefined/insurance/undefined/select-plan",
          "type": "router/NAVIGATE",
        }
      }
    />
  </div>
  <h1>
    Confirm an insurance plan
  </h1>
  <Connect(InsurancePlanDecorator)
    activating={false}
    findInsurancePlan={[MockFunction]}
    onConfirm={[Function]}
  />
</div>
`;
